src/main/java/org/swisspush/redisques/QueueStatsService.java

```diff
@@ -0,0 +1,219 @@
+        queueStatisticsCollector.getQueueStatistics(req.queueNames).onComplete( ev -> {
+            req.queueNames = null; // <- no longer needed
+            long durGetQueueStatsMs = currentTimeMillis() - begGetQueueStatsMs;
+            if (durGetQueueStatsMs > 42) log.debug("queueStatisticsCollector.getQueueStatistics() took {}ms", durGetQueueStatsMs);
+            if (ev.failed()) {
+                log.warn("queueStatisticsCollector.getQueueStatistics() failed. Fallback to empty result.", ev.cause());
+                req.queuesJsonArr = new JsonArray();
+                onDone.accept(null, req);
+                return;
+            }
+            JsonObject queStatsJsonObj = ev.result();
+            String status = queStatsJsonObj.getString(STATUS);
+            if (!OK.equals(status)) {
+                log.warn("queueStatisticsCollector.getQueueStatistics() responded '" + status + "'. Fallback to empty result.", ev.cause());
+                req.queuesJsonArr = new JsonArray();
+                onDone.accept(null, req);
+                return;
+            }
+            req.queuesJsonArr = queStatsJsonObj.getJsonArray(QUEUES);
+            onDone.accept(null, req);
+        });
+    }
+
+    private <CTX> void mergeRetryDetailsIntoCollectedData(GetQueueStatsRequest<CTX> req, BiConsumer<Throwable, GetQueueStatsRequest<CTX>> onDone) {
+        // Setup a lookup table as we need to find by name further below.
+        Map<String, JsonObject> detailsByName = new HashMap<>(req.queuesJsonArr.size());
+        for (var it = (Iterator<JsonObject>) (Object) req.queuesJsonArr.iterator(); it.hasNext(); ) {
+            JsonObject detailJson = it.next();
+            String name = detailJson.getString(MONITOR_QUEUE_NAME);
+            detailsByName.put(name, detailJson);
+        }
+        for (Queue queue : req.queues) {
+            JsonObject detail = detailsByName.get(queue.name);
+            if (detail == null) continue; // no details to enrich.
+            JsonObject dequeueStatsJson = detail.getJsonObject(STATISTIC_QUEUE_DEQUEUESTATISTIC);
+            if (dequeueStatsJson == null) continue; // no dequeue stats we could enrich
+            DequeueStatistic dequeueStats = dequeueStatsJson.mapTo(DequeueStatistic.class);
+            // Attach whatever details we got.
+            queue.lastDequeueAttemptEpochMs = dequeueStats.lastDequeueAttemptTimestamp;
+            queue.lastDequeueSuccessEpochMs = dequeueStats.lastDequeueSuccessTimestamp;
+            queue.nextDequeueDueTimestampEpochMs = dequeueStats.nextDequeueDueTimestamp;
+        }
+        onDone.accept(null, req);
+    }
+
+    private int compareLargestFirst(Queue aq, Queue bq) {
+        if (aq.size == null && bq.size == null) return 0;
+        if (aq.size == null) return -1;
+        if (bq.size == null) return +1;
+        long as = aq.size, bs = bq.size;
+        if (as > bs) return -1;
+        if (as < bs) return +1;
+        assert as == bs : as +", "+ bs;
```

Comment:
Is this assert really necessary? Something must have gone seriously wrong if you end up with `as != bs` here.

Reply:
That's exactly the point of assert.

Reply:
It's mainly confusing because you're already handling all possible outcomes above. There's basically no way this can happen apart from a glitch in the matrix! :-)